kyo-prelude/shared/src/main/scala/kyo2/kernel/Runtime.scala

```diff
@@ -0,0 +1,124 @@
+package kyo2.kernel
+
+import internal.*
+import java.util.Arrays
+import kyo2.isNull
+import scala.annotation.tailrec
+import scala.util.control.NonFatal
+
+final private class Runtime(initDepth: Int, initState: Runtime.State):
+    import Runtime.State
+
+    private val owner         = Thread.currentThread()
+    private[kernel] var depth = initDepth
+    private var traceIdx      = initState.traceIdx
+    private val trace         = Runtime.copyTrace(initState.trace, traceIdx)
+
+    private def enter(_frame: Frame, force: Boolean = false): Int =
+        assert(Thread.currentThread eq owner, "Leaked runtime! " + (Thread.currentThread, owner))
```

Comment:
This is something I'm not sure about: I'm using implicits as a way to propagate the current `Runtime` without having to read from `ThreadLocal` on each transformation to reduce the performance overhead. I tried to avoid it but there can be situations where a `Runtime` instance gets captured by a computation and reused later from another thread. I'll explore this further when porting fibers to the new design.

Reply:
I've made [a change](https://github.com/getkyo/kyo/pull/489/commits/9bb294f17b837ac76f12bcfed32eab7ac24a984c) to provide graceful handling of runtime leaks. In case of mismatch, the computation suspends and the corresponding `.eval` call will re-wire the correct `Runtime` instance for the current thread.

I've also added tests for it and a good thing is that I couldn't reproduce a leak using Kyo's APIs directly, only by manually saving the Runtime instance (last two scenarios).